[Merged by Bors] - feat(CategoryTheory): Exponentiable morphisms

sinhp opened the pull request:
We define an exponentiable morphism `f : I ⟶ J` to be a morphism with a chosen pullback functor
`Over J ⥤ Over I` together with a right adjoint, called the pushforward functor.

We use this PR in the PR branch `lccc-basic`.  

```mermaid
graph TD
  A[ChosenPullbacksAlong.lean] --> A'[Over.lean] 
  A' --> B[Sections.lean]
  A --> C[ExponentiableMorphism.lean]
  B --> D[Basic.lean]
  C --> D
  D --> E[Types.lean]
  E --> F[Presheaves.lean]
  D --> G[Beck-Chevalley.lean]

  %% Define highlight style
  classDef highlight fill:#ffe599,stroke:#d4a017,stroke-width:2px;

  %% Apply to one node
  class C highlight;
```

---
- [x] depends on: #31033
- [x] depends on: #31433

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sinhp committed:
ChosenPullback

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackObj

sinhp committed:
pullbackObj

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback fst

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback.hasPullbackAlong

sinhp committed:
ChosenPullback.pullback iso Over.pullback

sinhp committed:
remove second have

sinhp committed:
rm simp

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
rm have

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
naming

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
terminal simp

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm simp!

sinhp committed:
ChosenPullbacks C

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
HasPullbacks instance from ChosenPullbacks

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
remove overMkHom

sinhp committed:
doc string

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Create types.lean

sinhp committed:
rm types from this

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Create Types.lean

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Types.lean

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackIsoOverPullback_inv_app_

sinhp committed:
generalize pullbackMap

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
simp nf

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm mapPullbackDesc

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
proof wo convert

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
use (Over.forget ..).mapIso

sinhp committed:
ext instead of hom_ext

sinhp committed:
change fst to snd

this is because, as seen in lccc-section, we want to have
ChosenPullback.snd (toUnit X) (toUnit I) = CartesianMonoidalCategory.snd X I := rfl

sinhp committed:
doc string

sinhp committed:
naming

sinhp committed:
Create ExponentiableMorphism.lean

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean:
```diff
@@ -0,0 +1,179 @@
+  /-- The pushforward functor -/
+  pushforward : Over I ⥤ Over J
+  /-- The pushforward functor is right adjoint to the pullback functor -/
+  pullbackAdjPushforward (f) : pullback f ⊣ pushforward := by infer_instance
```
Adjunctions are not typeclasses:
```suggestion
  pullbackAdjPushforward (f) : pullback f ⊣ pushforward
```

sinhp committed:
computable cartesian monoidal for slices

sinhp committed:
pullbackId and pullbackComp

sinhp committed:
isoInv

sinhp committed:
add docstring

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
cartesian monoidal section moved to a new PR

sinhp committed:
remove namespace Over

sinhp committed:
computable cartesian monoidal structure on slices

sinhp committed:
exe mk_all

sinhp committed:
rename to ChosenPullbackAlong

sinhp committed:
improve docstrings

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

sinhp committed:
add s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
name change

sinhp committed:
Update Over.lean

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-over

sinhp committed:
sorry free

sinhp committed:
equivToOverUnit

sinhp committed:
Update Over.lean

sinhp committed:
remove simp from equality of functors

https://github.com/leanprover-community/mathlib4/pull/30373#discussion_r2526336070

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
binaryFan_fst & binaryFan_snd

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
explicit vars

sinhp committed:
Merge branch 'lccc-over' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
rm mixin lemmas

sinhp committed:
fix names

sinhp committed:
Update Mathlib.lean

sinhp committed:
namespace cleanup

dagurtomas committed:
Merge branch 'master' into lccc-over

dagurtomas committed:
Apply suggestion from @dagurtomas

dagurtomas committed:
move `@[expose] public section`

Expose the section for better accessibility.

sinhp committed:
Merge branch 'master' into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'lccc-over' into lccc-exp_mor

sinhp committed:
new module system

sinhp committed:
don't need Over

sinhp committed:
not needed in this PR

sinhp committed:
Update ExponentiableMorphism.lean

sinhp committed:
Update Mathlib.lean

sinhp committed:
Update Mathlib.lean

sinhp committed:
Update ExponentiableMorphism.lean

sinhp committed:
@[expose] public section

sinhp committed:
Merge branch 'master' into lccc-exp_mor

sinhp committed:
Delete Over.lean

sinhp committed:
Merge branch 'master' into lccc-exp_mor

sinhp committed:
fix

sinhp committed:
remove instance

sinhp committed:
generalize assumption for instances of id and comp

sinhp committed:
ev_coev fix

sinhp committed:
ev & coev def

sinhp commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean:
```diff
@@ -0,0 +1,164 @@
+
+/-- The dependent coevaluation natural transformation as the unit of the adjunction. -/
+abbrev coev : 𝟭 _ ⟶ pullback f ⋙ pushforward f :=
+  pullbackAdjPushforward f |>.unit
```
implemented this in the last commit.

sinhp committed:
fix

sinhp committed:
also for comp

sinhp committed:
rm simp

sinhp committed:
fix linter simp errors

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean:
```diff
@@ -0,0 +1,180 @@
+  pullbackAdjPushforward f |>.left_triangle_components X
+
+/-- The second triangle identity for the counit and unit of the adjunction. -/
+@[reassoc]
```
```suggestion
@[reassoc (attr := simp)]
```

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean:
```diff
@@ -0,0 +1,180 @@
+def pushforwardId (I : C) [ChosenPullbacksAlong (𝟙 I)] [ExponentiableMorphism (𝟙 I)] :
+    pushforward (𝟙 I) ≅ 𝟭 (Over I) :=
+  Adjunction.rightAdjointUniq (pullbackAdjPushforward (𝟙 I)) (id I).pullbackAdjPushforward
+
```
I think it might be useful to have the same simp lemmas as for `ChosenPullbacks` that characterizes this isomorphism (e.g `unit_pushforwardId_hom_app` etc.), and same for `comp`

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update ExponentiableMorphism.lean

sinhp commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean:
```diff
@@ -0,0 +1,180 @@
+  coev f |>.naturality g
+
+/-- The first triangle identity for the counit and unit of the adjunction. -/
+@[reassoc]
```
Adding simps unfortunately result in linter errors since `simp` could prove these two lemmas.

sinhp committed:
Merge branch 'master' into lccc-exp_mor

sinhp committed:
for consistency

robin-carlier reviewed (commented):
Thanks!

maintainer merge

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean:
```diff
@@ -0,0 +1,220 @@
+    (pullback f).map ((pushforward f).map g) ≫ (ev f).app Y = (ev f).app X ≫ g :=
+  ev f |>.naturality g
+
+theorem coev_naturality {X Y : Over J} (g : X ⟶ Y) :
```
```suggestion
@[reassoc]
theorem coev_naturality {X Y : Over J} (g : X ⟶ Y) :
```

joelriou committed:
Apply suggestion from @joelriou

joelriou committed:
Apply suggestion from @joelriou
